src/moves.rs

```diff
@@ -53,9 +75,12 @@ impl Board {
                 }
 
                 hands.use_tile(player, tile)?;
-                self.set(position, player, tile)?;
-                self.resolve_attack(player, position, judge);
-                Ok(())
+                changes.push(Change {
+                    detail: self.set(position, player, tile)?,
```

Comment:
Could be nice (in the future) to just collect changes and apply mutations at a higher level. Would prevent weird states if there's an error after mutation, and would help separate concerns.

Reply:
Yeah exactly where all of this lives is a good question. For now I was just writing it in spots where I had to bubble the least detail upwards — it would be good to do a holistic refactor of it in the future